Platform accessibility text scaling

TomEdwardsEnscape opened the pull request:
Many platforms support text scaling, an accessibility feature which allows the user to request that text be drawn larger (or on some platforms smaller) than normal, without altering other UI elements. Avalonia currently has no support for this feature. This PR adds support for iOS and Windows, and creates the infrastructure necessary to extend support to more platforms in the future.

The feature is disabled by default, so there are no breaking changes. It has been enabled for the Control Gallery.

Only `TextBlock`, `Inline`, and `TextPresenter` scale text by themselves. There is no change to low-level text measurement or rendering.

## What is the updated/expected behavior with this PR?

The Control Gallery now looks like this with 150% text scaling on Windows:

<img width="770" height="623" alt="Screenshot 2025-12-16 130913" src="https://github.com/user-attachments/assets/d63f5155-af12-4b38-b86b-cb714a73ae97" />

Note that the window, tiles, scrollbars and other non-text elements are the same size as before. Only the text is larger.

## How was the solution implemented (if it's not obvious)?

Text scaling is provided via `IPlatformSettings`. `TextBlock`, `Inline`, and `TextPresenter` get the platform settings object from their `TopLevel` and ask it to adjust their assigned `FontSize` when they are building their text layout objects. The value of `FontSize` does not change; platform text scaling is an extra layer on top which affects only measurement and rendering. This is how WinUI3 handles the feature.

Each individual control can configure the inherited `TextElement.IsPlatformTextScalingEnabled` property to control whether text scaling is applied to it and its children. Inline elements use the value of their host.

## Breaking changes
None.

## Obsoletions / Deprecations
None.

TomEdwardsEnscape committed:
Added support for platform text scaling, disabled by default
Only TextBlock, Inline, and TextPresenter scale text by themselves
No change to low-level text rendering or FontSize values
Implemented for iOS and Windows
Windows scaling is currently uniform regardless of base font size, which is incorrect

maxkatz6 commented on src/Windows/Avalonia.Win32/Win32PlatformSettings.cs:
```diff
@@ -87,5 +98,18 @@ internal void OnColorValuesChanged()
+
+    public override double GetScaledFontSize(double baseFontSize)
+    {
+#warning larger text should scale less, but the algorithm for this is closed source and undocumented
```
https://github.com/microsoft/microsoft-ui-xaml/blob/5788dee271452753d5b2c70179f976c3e96a45c7/src/dxaml/xcp/core/text/common/TextFormatting.cpp#L181-L204

maxkatz6 replied:
It also seems character/letter spacing should also be scaled https://github.com/microsoft/microsoft-ui-xaml/blob/5788dee271452753d5b2c70179f976c3e96a45c7/src/dxaml/xcp/core/native/text/Controls/TextBox.cpp#L793

TomEdwardsEnscape committed:
Also scale LineHeight, LineSpacing, and LetterSpacing
Fixed TextBox.MaxLines and TextBox.MinLines calculations when text scaling is active
Fixed TextBox not updating when LineSpacing changes
Removed EffectiveFontSize, since it's no longer the only property which scales

TomEdwardsEnscape committed:
Corrected iOS font size scaling units

TomEdwardsEnscape commented:
> To be more generic, we'd like the text scale factor to be configurable and inherited, with its default coming from the platform. Proposed name: `TextScaleFactor`. This should replace `IsPlatformTextScalingEnabled`.

This can't be done. Text scaling is not a factor but a per-platform algorithm. On Windows this algorthim is a complicated mathematical function, while on iOS it's an absolute point increase or decrease, plus a minimum size. Android may do something else.

> We're fine with making the platform scaling the default in v12. This PR won't be backported to v11.

Alright, but my iOS project targets v11, is due for release in Q1 next year, and we would like it to support text scaling. We'll probably make a cheeky custom build by opening a backport PR despite this decision.

TomEdwardsEnscape committed:
Added TextScaling static class with configuration properties
Added ITextScaler to support custom text scaling
Calculate a font scale factor and use this to adjust height and spacing values
Fixed TextBox not updating when LineHeight changes and MaxLines or MinLines are active

TomEdwardsEnscape committed:
Merged master

TomEdwardsEnscape committed:
Fix _fontScaleCache not ever being populated

TomEdwardsEnscape committed:
Merged master

TomEdwardsEnscape committed:
Merged master

Gillibald commented:
I had a deeper look at this PR, and I think we should move the scaling logic to a different layer. At the moment, the scaling is applied at different places to different values that are used for text layout. Instead, we should make the scaling a property of text layout. Therefore, I propose introducing the following addition to TextRunProperties and TextParagraphProperties:
```
public abstract class TextRunProperties
{
    /// <summary>
    /// Scale applied to <see cref="FontRenderingEmSize"/> to obtain the em size the run is
    /// actually formatted and rendered at. 1 by default.
    /// </summary>
    public virtual double FontSizeScale => 1;
}

public abstract class TextParagraphProperties
{
    /// <summary>
    /// Scale applied to the size-derived paragraph metrics: <see cref="LineHeight"/>,
    /// LineSpacing, <see cref="LetterSpacing"/>, <see cref="Indent"/>,
    /// <see cref="ParagraphIndent"/> and <see cref="DefaultIncrementalTab"/>. 1 by default.
    /// </summary>
    public virtual double FontSizeScale => 1;
}
```
The scaling is then synced as any other value. So in the end, the ITextScaler needs to produce a scaling factor instead of scaling values individually. We can't scale parts of the text metrics individually. Everything needs to be scaled uniformly.

This change makes sure all consumers of FormattedText and TextLayout can apply the current scaling.

The final FontSizeScale is applied where we already apply the em-size, so all measurements reflect the scaled values.

Gillibald commented on src/Avalonia.Controls/TopLevel.cs:
```diff
@@ -241,6 +242,8 @@ internal TopLevel(ITopLevelImpl impl, IAvaloniaDependencyResolver? dependencyRes
 
             CreatePlatformImplBinding(TransparencyLevelHintProperty, hint => PlatformImpl.SetTransparencyLevelHint(hint ?? Array.Empty<WindowTransparencyLevel>()));
 
+            PlatformSettings?.TextScalingChanged += OnPlatformTextScalingChanged;
```
This is leaking because we never unsubscribe

Gillibald commented on src/Windows/Avalonia.Win32/Win32PlatformSettings.cs:
```diff
@@ -101,6 +111,13 @@ internal void OnColorValuesChanged()
+        if (newTextScaleFactor != _textScaleFactor)
+        {
+            _textScaleFactor = newTextScaleFactor;
+            OnTextScaleChanged();
```
TextScaling is never updated on its own. Only a color change triggers the update. We should listen to text scaling changes explicitly.

TomEdwardsEnscape replied:
There is the method `IUISettings2.AddTextScaleFactorChanged` which is supposed to allow us to do this, but I was unable to get it to work. The signature is funky, as it requires a `void*` parameter.

I'm not aware of any other direct way to receive notifications of text scaling changes. There doesn't appear to be a window message.

Gillibald commented on src/Avalonia.Controls/TextScaling.cs:
```diff
@@ -0,0 +1,140 @@
+/// <seealso cref="ITextScaler"/>
+public static class TextScaling
+{
+    private static readonly ConditionalWeakTable<ITextScaler, HashSet<Visual>> s_customTextScalerSubscribers = [];
```
This leaks visuals. The value holds strong Visual refs, and nothing removes them on detach, so any custom scaler that outlives the visuals using it (a singleton in App.Resources, the normal case) roots the whole subtree.

TomEdwardsEnscape commented:
> We should make the scaler itself a value that gets inherited with the following design:

This is already how custom scalers are set up, so the difference in this proposal is that the standard platform scaling is now also an inherited, per-element property.

I have the following questions about this idea:

1. What value should the user provide if they want ordinary per-platform text scaling? Your sample code doesn't define this. We can define a static scaler called "Platform" maybe? This works but is harder to use than `TextScaling.IsEnabled="True"`.

2. What are the performance implications? The common case will be for a platform scaler object to be assigned to the `TopLevel`. This is a nondefault value, so must propagate up and down the entire visual tree, raise events, and so on. The current design skips all this book-keeping by going straight up to the visual tree root unless a custom scaler is assigned.

3. What happened to the min/max font size properties currently defined in `TextScaling`?

> This design allows every consumer of TextRunProperties and TextParagraphProperties to properly scale values that are derived from the em size

I avoided doing this because not all consumers of those types will want to deal with text scaling, and it's not trivial to get everything right.

If you are drawing a custom text layout you can't just slap a scaling value on and forget about it. You also have to think about what impact this has on the bounds of the text. You may need to move or scale images and other graphics to match. And then you may have to clear caches of this or that (e.g. bitmaps containing previous render output) when text scaling changes.

I foresee lots of buggy implementations here, so I think it's better not to set developers up for failure by making it easy to pass a scaler in "because it's there" without understanding its implications. Text scaling applied by modifying the existing font size, line height, etc. values is a bit more typing but has the benefit of requiring the developer to understand what they are doing.

Gillibald commented:
Every other property of TextElement is synced into TextRunProperties, etc., and that is properly invalidating the rendering and layout. I don't understand why that would be an issue. TextScaling affects only text, not any other content. The default TextScaler calls into a platform implementation. I want to avoid attaching visuals to a text scaler, so I chose this design.
